File: app/src/main/java/com/tunes/player/utils/PlaylistStorageManager.java
package com.tunes.player.utils;

import android.content.Context;
import android.util.Log;

import com.tunes.player.R;
import com.tunes.player.model.MusicModel;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Stores playlist, history, favourites and imported-track data as JSON inside
 * the app's private files directory.  JSON replaces the previous Java-serialization
 * approach (which had no type-filter and could deserialise unexpected classes).
 */
public class PlaylistStorageManager {

    private static final String TAG = "PlaylistStorageManager";

    private static final String FOLDER_HISTORY       = "history";
    private static final String PLAYLIST_DATA_FILE   = "Playlist_Tracks_Data.json";
    private static final String PLAYLIST_TITLES_FILE = "Playlist_Names.json";
    private static final String FAVORITE_TRACKS_FILE = "FavoriteTracks.json";
    private static final String IMPORTED_TRACKS_FILE = "ImportedTracks.json";

    private PlaylistStorageManager() {}

    // -------------------------------------------------------------------------
    // History
    // -------------------------------------------------------------------------

    public static void addToRecentTracks(Context context, MusicModel md) {
        new Thread(() -> {
            File dir = new File(context.getFilesDir(), FOLDER_HISTORY);
            if (!dir.exists()) dir.mkdirs();
            String safeName = md.getSongName()
                    .replaceAll("[\\\\/:*?\"<>|]", "_") + ".json";
            File f = new File(dir, safeName);
            try {
                writeJson(f, musicModelToJson(md).toString());
            } catch (JSONException | IOException e) {
                Log.e(TAG, "addToRecentTracks failed", e);
            }
        }).start();
    }

    public static List<MusicModel> getRecentTracks(Context context) {
        File dir = new File(context.getFilesDir(), FOLDER_HISTORY);
        File[] files = dir.listFiles();
        List<MusicModel> list = new ArrayList<>();
        if (files == null) return list;

        Arrays.sort(files, (o1, o2) -> Long.compare(o2.lastModified(), o1.lastModified()));
        int count = 0;
        for (File f : files) {
            if (count >= 50) break;
            try {
                String json = readFile(f);
                if (json != null) {
                    list.add(musicModelFromJson(new JSONObject(json)));
                    count++;
                }
            } catch (JSONException | IOException e) {
                Log.e(TAG, "getRecentTracks error reading " + f.getName(), e);
            }
        }
        return list;
    }

    // -------------------------------------------------------------------------
    // Favourites
    // -------------------------------------------------------------------------

    public static void saveFavorite(Context context, List<MusicModel> list) {
        saveTrackList(context, FAVORITE_TRACKS_FILE, list);
    }

    public static List<MusicModel> getFavorite(Context context) {
        return loadTrackList(context, FAVORITE_TRACKS_FILE);
    }

    // -------------------------------------------------------------------------
    // Imported tracks
    // -------------------------------------------------------------------------

    public static void saveImportedTracks(Context context, List<MusicModel> list) {
        new Thread(() -> {
            synchronized (IMPORTED_TRACKS_FILE) {
                List<MusicModel> current = loadTrackList(context, IMPORTED_TRACKS_FILE);
                boolean modified = false;
                for (MusicModel m : list) {
                    boolean exists = false;
                    for (MusicModel c : current) {
                        if (c.getSongPath() != null && c.getSongPath().equals(m.getSongPath())) {
                            exists = true;
                            break;
                        }
                    }
                    if (!exists) {
                        current.add(m);
                        modified = true;
                    }
                }
                if (modified) saveTrackListSync(context, IMPORTED_TRACKS_FILE, current);
            }
        }).start();
    }

    public static List<MusicModel> getImportedTracks(Context context) {
        return loadTrackList(context, IMPORTED_TRACKS_FILE);
    }

    // -------------------------------------------------------------------------
    // Playlist titles
    // -------------------------------------------------------------------------

    public static void savePlaylistTitles(Context context, List<String> titles) {
        new Thread(() -> {
            JSONArray arr = new JSONArray();
            for (String t : titles) arr.put(t);
            File f = new File(context.getFilesDir(), PLAYLIST_TITLES_FILE);
            try {
                writeJson(f, arr.toString());
            } catch (IOException e) {
                Log.e(TAG, "savePlaylistTitles failed", e);
            }
        }).start();
    }

    public static List<String> getPlaylistTitles(Context context) {
        List<String> list = new ArrayList<>();
        File f = new File(context.getFilesDir(), PLAYLIST_TITLES_FILE);
        if (!f.exists()) {
            list.add(context.getString(R.string.playlist_current_queue));
            list.add(context.getString(R.string.favorite_playlist));
            return list;
        }
        try {
            String json = readFile(f);
            if (json != null) {
                JSONArray arr = new JSONArray(json);
                for (int i = 0; i < arr.length(); i++) list.add(arr.getString(i));
            }
        } catch (JSONException | IOException e) {
            Log.e(TAG, "getPlaylistTitles failed", e);
        }
        if (list.isEmpty()) {
            list.add(context.getString(R.string.playlist_current_queue));
            list.add(context.getString(R.string.favorite_playlist));
        }
        return list;
    }

    // -------------------------------------------------------------------------
    // Playlist tracks
    // -------------------------------------------------------------------------

    public static void updatePlaylistTracks(Context context, List<MusicModel> updatedList, int index) {
        new Thread(() -> {
            synchronized (PLAYLIST_DATA_FILE) {
                List<List<MusicModel>> all = loadAllPlaylistTracks(context);
                while (all.size() <= index) all.add(new ArrayList<>());
                all.set(index, updatedList != null ? new ArrayList<>(updatedList) : new ArrayList<>());
                saveAllPlaylistTracks(context, all);
            }
        }).start();
    }

    public static List<MusicModel> getPlaylistTrackAtPosition(Context context, int pos) {
        List<List<MusicModel>> all = loadAllPlaylistTracks(context);
        if (pos >= 0 && pos < all.size()) return new ArrayList<>(all.get(pos));
        return new ArrayList<>();
    }

    public static void dropPlaylistCardDataAt(Context context, int pos) {
        new Thread(() -> {
            synchronized (PLAYLIST_DATA_FILE) {
                int customIndex = pos - 2;
                if (customIndex < 0) return;
                List<List<MusicModel>> all = loadAllPlaylistTracks(context);
                if (customIndex < all.size()) {
                    all.remove(customIndex);
                    saveAllPlaylistTracks(context, all);
                }
            }
        }).start();
    }

    public static void dropAllPlaylistData(Context context) {
        new Thread(() -> {
            new File(context.getFilesDir(), PLAYLIST_TITLES_FILE).delete();
            new File(context.getFilesDir(), PLAYLIST_DATA_FILE).delete();
        }).start();
    }

    // -------------------------------------------------------------------------
    // Private helpers
    // -------------------------------------------------------------------------

    private static void saveTrackList(Context context, String fileName, List<MusicModel> list) {
        new Thread(() -> saveTrackListSync(context, fileName, list)).start();
    }

    private static synchronized void saveTrackListSync(Context context, String fileName, List<MusicModel> list) {
        File f = new File(context.getFilesDir(), fileName);
        if (list == null || list.isEmpty()) {
            if (f.exists() && f.delete()) Log.d(TAG, fileName + " deleted");
            return;
        }
        JSONArray arr = new JSONArray();
        for (MusicModel m : list) {
            try { arr.put(musicModelToJson(m)); } catch (JSONException e) {
                Log.e(TAG, "saveTrackListSync serialise error", e);
            }
        }
        try {
            writeJson(f, arr.toString());
        } catch (IOException e) {
            Log.e(TAG, "saveTrackListSync write error for " + fileName, e);
        }
    }

    private static synchronized List<MusicModel> loadTrackList(Context context, String fileName) {
        List<MusicModel> list = new ArrayList<>();
        File f = new File(context.getFilesDir(), fileName);
        if (!f.exists()) return list;
        try {
            String json = readFile(f);
            if (json == null) return list;
            JSONArray arr = new JSONArray(json);
            for (int i = 0; i < arr.length(); i++) {
                list.add(musicModelFromJson(arr.getJSONObject(i)));
            }
        } catch (JSONException | IOException e) {
            Log.e(TAG, "loadTrackList error for " + fileName, e);
        }
        return list;
    }

    private static synchronized void saveAllPlaylistTracks(Context context, List<List<MusicModel>> listOfLists) {
        File f = new File(context.getFilesDir(), PLAYLIST_DATA_FILE);
        JSONArray outer = new JSONArray();
        for (List<MusicModel> inner : listOfLists) {
            JSONArray arr = new JSONArray();
            for (MusicModel m : inner) {
                try { arr.put(musicModelToJson(m)); } catch (JSONException e) {
                    Log.e(TAG, "saveAllPlaylistTracks serialise error", e);
                }
            }
            outer.put(arr);
        }
        try {
            writeJson(f, outer.toString());
        } catch (IOException e) {
            Log.e(TAG, "saveAllPlaylistTracks write error", e);
        }
    }

    private static synchronized List<List<MusicModel>> loadAllPlaylistTracks(Context context) {
        List<List<MusicModel>> result = new ArrayList<>();
        File f = new File(context.getFilesDir(), PLAYLIST_DATA_FILE);
        if (!f.exists()) return result;
        try {
            String json = readFile(f);
            if (json == null) return result;
            JSONArray outer = new JSONArray(json);
            for (int i = 0; i < outer.length(); i++) {
                JSONArray inner = outer.getJSONArray(i);
                List<MusicModel> list = new ArrayList<>();
                for (int j = 0; j < inner.length(); j++) {
                    list.add(musicModelFromJson(inner.getJSONObject(j)));
                }
                result.add(list);
            }
        } catch (JSONException | IOException e) {
            Log.e(TAG, "loadAllPlaylistTracks error", e);
        }
        return result;
    }

    // -------------------------------------------------------------------------
    // JSON <-> MusicModel
    // -------------------------------------------------------------------------

    private static JSONObject musicModelToJson(MusicModel m) throws JSONException {
        JSONObject o = new JSONObject();
        o.put("id",        m.getId());
        o.put("name",      m.getSongName());
        o.put("artist",    m.getArtist());
        o.put("album",     m.getAlbum());
        o.put("path",      m.getSongPath()    != null ? m.getSongPath()    : "");
        o.put("art",       m.getAlbumArtUrl() != null ? m.getAlbumArtUrl() : "");
        o.put("duration",  m.getDuration());
        o.put("dateAdded", m.getDateAdded());
        return o;
    }

    private static MusicModel musicModelFromJson(JSONObject o) throws JSONException {
        return new MusicModel(
                o.getInt("id"),
                o.optString("name",    ""),
                o.optString("artist",  ""),
                o.optString("album",   ""),
                o.optString("path",    null),
                o.optString("art",     null),
                o.optLong("duration",  0),
                o.optLong("dateAdded", 0)
        );
    }

    // -------------------------------------------------------------------------
    // File I/O
    // -------------------------------------------------------------------------

    private static void writeJson(File file, String json) throws IOException {
        try (BufferedWriter w = new BufferedWriter(new FileWriter(file, false))) {
            w.write(json);
        }
    }

    private static String readFile(File file) throws IOException {
        StringBuilder sb = new StringBuilder();
        try (BufferedReader r = new BufferedReader(new FileReader(file))) {
            String line;
            while ((line = r.readLine()) != null) sb.append(line);
        }
        return sb.length() > 0 ? sb.toString() : null;
    }
}

File: app/src/main/java/com/jegly/frequency/utils/PlaylistStorageManager.kt
package com.jegly.frequency.utils

import android.content.Context
import android.util.Log
import com.jegly.frequency.R
import com.jegly.frequency.model.MusicModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.io.*

object PlaylistStorageManager {

    private const val TAG = "PlaylistStorageManager"
    private const val FOLDER_HISTORY = "history"
    private const val PLAYLIST_DATA_FILE = "Playlist_Tracks_Data.json"
    private const val PLAYLIST_TITLES_FILE = "Playlist_Names.json"
    private const val FAVORITE_TRACKS_FILE = "FavoriteTracks.json"
    private const val IMPORTED_TRACKS_FILE = "ImportedTracks.json"

    suspend fun addToRecentTracks(context: Context, md: MusicModel) = withContext(Dispatchers.IO) {
        val dir = File(context.filesDir, FOLDER_HISTORY)
        if (!dir.exists()) dir.mkdirs()
        val safeName = md.songName.replace("[\\\\/:*?\"<>|]".toRegex(), "_") + ".json"
        val f = File(dir, safeName)
        try {
            writeJson(f, musicModelToJson(md).toString())
        } catch (e: Exception) {
            Log.e(TAG, "addToRecentTracks failed", e)
        }
    }

    suspend fun getRecentTracks(context: Context): List<MusicModel> = withContext(Dispatchers.IO) {
        val dir = File(context.filesDir, FOLDER_HISTORY)
        val files = dir.listFiles()
        val list = mutableListOf<MusicModel>()
        if (files == null) return@withContext list

        files.sortByDescending { it.lastModified() }
        var count = 0
        for (f in files) {
            if (count >= 50) break
            try {
                val json = readFile(f)
                if (json != null) {
                    list.add(musicModelFromJson(JSONObject(json)))
                    count++
                }
            } catch (e: Exception) {
                Log.e(TAG, "getRecentTracks error reading ${f.name}", e)
            }
        }
        list
    }

    suspend fun saveFavorite(context: Context, list: List<MusicModel>) {
        saveTrackList(context, FAVORITE_TRACKS_FILE, list)
    }

    suspend fun getFavorite(context: Context): List<MusicModel> = withContext(Dispatchers.IO) {
        loadTrackList(context, FAVORITE_TRACKS_FILE)
    }

    suspend fun saveImportedTracks(context: Context, list: List<MusicModel>) = withContext(Dispatchers.IO) {
        synchronized(IMPORTED_TRACKS_FILE) {
            val current = loadTrackList(context, IMPORTED_TRACKS_FILE).toMutableList()
            var modified = false
            for (m in list) {
                val exists = current.any { it.songPath == m.songPath }
                if (!exists) {
                    current.add(m)
                    modified = true
                }
            }
            if (modified) saveTrackListSync(context, IMPORTED_TRACKS_FILE, current)
        }
    }

    suspend fun getImportedTracks(context: Context): List<MusicModel> = withContext(Dispatchers.IO) {
        loadTrackList(context, IMPORTED_TRACKS_FILE)
    }

    suspend fun savePlaylistTitles(context: Context, titles: List<String>) = withContext(Dispatchers.IO) {
        val arr = JSONArray()
        for (t in titles) arr.put(t)
        val f = File(context.filesDir, PLAYLIST_TITLES_FILE)
        try {
            writeJson(f, arr.toString())
        } catch (e: IOException) {
            Log.e(TAG, "savePlaylistTitles failed", e)
        }
    }

    suspend fun getPlaylistTitles(context: Context): List<String> = withContext(Dispatchers.IO) {
        val list = mutableListOf<String>()
        val f = File(context.filesDir, PLAYLIST_TITLES_FILE)
        if (!f.exists()) {
            return@withContext listOf(
                context.getString(R.string.playlist_current_queue),
                context.getString(R.string.favorite_playlist)
            )
        }
        try {
            val json = readFile(f)
            if (json != null) {
                val arr = JSONArray(json)
                for (i in 0 until arr.length()) list.add(arr.getString(i))
            }
        } catch (e: Exception) {
            Log.e(TAG, "getPlaylistTitles failed", e)
        }
        if (list.isEmpty()) {
            listOf(
                context.getString(R.string.playlist_current_queue),
                context.getString(R.string.favorite_playlist)
            )
        } else list
    }

    suspend fun updatePlaylistTracks(context: Context, updatedList: List<MusicModel>?, index: Int) = withContext(Dispatchers.IO) {
        synchronized(PLAYLIST_DATA_FILE) {
            val all = loadAllPlaylistTracks(context).toMutableList()
            while (all.size <= index) all.add(mutableListOf())
            all[index] = updatedList?.toMutableList() ?: mutableListOf()
            saveAllPlaylistTracks(context, all)
        }
    }

    suspend fun getPlaylistTrackAtPosition(context: Context, pos: Int): List<MusicModel> = withContext(Dispatchers.IO) {
        val all = loadAllPlaylistTracks(context)
        if (pos in all.indices) all[pos].toList() else emptyList()
    }

    suspend fun dropPlaylistCardDataAt(context: Context, pos: Int) = withContext(Dispatchers.IO) {
        synchronized(PLAYLIST_DATA_FILE) {
            val customIndex = pos - 2
            if (customIndex < 0) return@synchronized
            val all = loadAllPlaylistTracks(context).toMutableList()
            if (customIndex in all.indices) {
                all.removeAt(customIndex)
                saveAllPlaylistTracks(context, all)
            }
        }
    }

    suspend fun dropAllPlaylistData(context: Context) = withContext(Dispatchers.IO) {
        File(context.filesDir, PLAYLIST_TITLES_FILE).delete()
        File(context.filesDir, PLAYLIST_DATA_FILE).delete()
    }

    private suspend fun saveTrackList(context: Context, fileName: String, list: List<MusicModel>) = withContext(Dispatchers.IO) {
        saveTrackListSync(context, fileName, list)
    }

    private fun saveTrackListSync(context: Context, fileName: String, list: List<MusicModel>) {
        val f = File(context.filesDir, fileName)
        if (list.isEmpty()) {
            if (f.exists() && f.delete()) Log.d(TAG, "$fileName deleted")
            return
        }
        val arr = JSONArray()
        for (m in list) {
            try {
                arr.put(musicModelToJson(m))
            } catch (e: JSONException) {
                Log.e(TAG, "saveTrackListSync serialise error", e)
            }
        }
        try {
            writeJson(f, arr.toString())
        } catch (e: IOException) {
            Log.e(TAG, "saveTrackListSync write error for $fileName", e)
        }
    }

    private fun loadTrackList(context: Context, fileName: String): List<MusicModel> {
        val list = mutableListOf<MusicModel>()
        val f = File(context.filesDir, fileName)
        if (!f.exists()) return list
        try {
            val json = readFile(f) ?: return list
            val arr = JSONArray(json)
            for (i in 0 until arr.length()) {
                list.add(musicModelFromJson(arr.getJSONObject(i)))
            }
        } catch (e: Exception) {
            Log.e(TAG, "loadTrackList error for $fileName", e)
        }
        return list
    }

    private fun saveAllPlaylistTracks(context: Context, listOfLists: List<List<MusicModel>>) {
        val f = File(context.filesDir, PLAYLIST_DATA_FILE)
        val outer = JSONArray()
        for (inner in listOfLists) {
            val arr = JSONArray()
            for (m in inner) {
                try {
                    arr.put(musicModelToJson(m))
                } catch (e: JSONException) {
                    Log.e(TAG, "saveAllPlaylistTracks serialise error", e)
                }
            }
            outer.put(arr)
        }
        try {
            writeJson(f, outer.toString())
        } catch (e: IOException) {
            Log.e(TAG, "saveAllPlaylistTracks write error", e)
        }
    }

    private fun loadAllPlaylistTracks(context: Context): List<List<MusicModel>> {
        val result = mutableListOf<List<MusicModel>>()
        val f = File(context.filesDir, PLAYLIST_DATA_FILE)
        if (!f.exists()) return result
        try {
            val json = readFile(f) ?: return result
            val outer = JSONArray(json)
            for (i in 0 until outer.length()) {
                val inner = outer.getJSONArray(i)
                val list = mutableListOf<MusicModel>()
                for (j in 0 until inner.length()) {
                    list.add(musicModelFromJson(inner.getJSONObject(j)))
                }
                result.add(list)
            }
        } catch (e: Exception) {
            Log.e(TAG, "loadAllPlaylistTracks error", e)
        }
        return result
    }

    private fun musicModelToJson(m: MusicModel): JSONObject {
        val o = JSONObject()
        o.put("id", m.id)
        o.put("name", m.songName)
        o.put("artist", m.artist)
        o.put("album", m.album)
        o.put("path", m.songPath ?: "")
        o.put("art", m.albumArtUrl ?: "")
        o.put("duration", m.duration)
        o.put("dateAdded", m.dateAdded)
        return o
    }

    private fun musicModelFromJson(o: JSONObject): MusicModel {
        return MusicModel(
            o.getLong("id"),
            o.optString("name", ""),
            o.optString("artist", ""),
            o.optString("album", ""),
            o.optString("path", null),
            o.optString("art", null),
            o.optLong("duration", 0),
            o.optLong("dateAdded", 0)
        )
    }

    private fun writeJson(file: File, json: String) {
        BufferedWriter(FileWriter(file, false)).use { it.write(json) }
    }

    private fun readFile(file: File): String? {
        val sb = StringBuilder()
        BufferedReader(FileReader(file)).use { r ->
            var line: String?
            while (r.readLine().also { line = it } != null) sb.append(line)
        }
        return if (sb.isNotEmpty()) sb.toString() else null
    }
}
